temba/api/v2/views.py

```diff
@@ -4924,44 +4946,89 @@ class MessagesReportEndpoint(BaseAPIView, ReportEndpointMixin):
+        ).only("flow_id", "events", "modified_on")
+        self.update_paginator_conf_for_flow_runs(runs)
+        runs, next_page = self.get_paginated_queryset(runs)
+        messages_uuids = [
+            evt["msg"].get("uuid")
+            for run in runs
+            for evt in run.get_msg_events()
+            if evt["msg"].get("uuid") not in [None, "None", ""]
+        ]
+        self.attach_messages_uuids_join(qs, messages_uuids)
+        qs = qs.only("created_on").annotate(ds=Concat("direction", "status"))
+        return qs, next_page
+
+    def update_paginator_conf_for_flow_runs(self, flow_run_qs):
+        run = flow_run_qs.first()
+        msgs_count = len([1 for evt in run.get_msg_events() if evt["msg"].get("uuid") not in [None, "None", ""]])
+        page_size = max(self.get_page_size() // msgs_count, 1)
+        setattr(self, "chunk_size", page_size)
+        setattr(self, "pagination_class", ModifiedOnCursorPagination)
+        setattr(self, "page_size_query_param", "__removed_page_size")
+
+    def get_page_size(self):
+        page_size = self.request.query_params.get("page_size") or 2000
+        try:
+            return int(page_size)
+        except ValueError:
+            return 2000
+
+    @staticmethod
+    def attach_messages_uuids_join(queryset: QuerySet, uuids: list):
+        msgs_uuid_pairs = ",".join(str((seq, model_uuid)) for seq, model_uuid in enumerate(uuids, start=1))
+
+        class MessagesUuidJoin:
+            table_name = "t"
+            join_type = "INNER JOIN"
+            parent_alias = "msgs_msg"
+            filtered_relation = None
+            nullable = None
+
+            @classmethod
+            def as_sql(cls, *args):
+                return f"INNER JOIN (VALUES {msgs_uuid_pairs}) t (seq, uuid) ON msgs_msg.uuid = t.uuid::uuid", []
+
+        queryset.query.join(MessagesUuidJoin)
 
     @csv_response_wrapper
     def get(self, request, *args, **kwargs):
         org = self.request.user.get_org()
         queryset = Msg.objects.filter(org=org)
         self.applied_filters, arg_filters = {}, []
 
-        flow = self.get_query_parameter("flow")
-        queryset = self.get_flow_messages(org, flow, queryset) if flow else queryset
         arg_filters.append(self.get_name_uuid_filters("exclude", "contact__all_groups", exclude=True))
         arg_filters.append(self.get_name_uuid_filters("channel", "channel"))
         arg_filters.append(self.get_datetime_filters("", "created_on", org))
         arg_filters = [_filter for _filter in arg_filters if _filter]
+        queryset = queryset.filter(*arg_filters)
+
+        try:
+            flow__uuid = self.get_query_parameter("flow")
+            if flow__uuid:
+                flow = Flow.objects.get(org=org, uuid=flow__uuid)
+                current_page, next_page = self.get_flow_messages(org, flow, queryset)
+            else:
+                queryset = queryset.only("created_on").annotate(ds=Concat("direction", "status"))
```

Comment:
@ihor-palii  - The issue that we had was exactly on this query, so to make our lives easier, let's make the "flow" query parameter required.

Reply:
Updated